File: solutions/TypeScript/[0121_Easy] Best Time to Buy and Sell Stock/maxProfit.ts
function maxProfit(prices: number[]): number {
  let minPrice = prices[0];
  let possibleMaxProfit = 0;

  for (let i = 0; i < prices.length; i++) {
    if (prices[i] < minPrice) {
      minPrice = prices[i];
    } else if (prices[i] - minPrice > possibleMaxProfit) {
      possibleMaxProfit = prices[i] - minPrice;
    }
  }

  return possibleMaxProfit;
}

File: solutions/Java/[0714_Medium] Best Time to Buy and Sell Stock with Transaction Fee/maxProfit.java
class Solution {
  public int maxProfit(int[] prices, int fee) {
    int n = prices.length;
    int free = 0;
    int hold = -prices[0];

    for (int i = 1; i < n; i++) {
      int previousHold = hold;
      hold = Math.max(hold, free - prices[i]);
      free = Math.max(free, previousHold + prices[i] - fee);
    }

    return free;
  }
}
